### PR TITLE
feat(CycloneDX): Add new agent cyclonedx

### PR DESCRIPTION
**Pull request by its-sushant**
## Description

This pull request introduces a new agent, CycloneDX, responsible for generating a software bill of materials (SBOM) in the CycloneDX format.

## How to test

To test the functionality of the CycloneDX agent in Fossology, you can follow these steps:

-  Install FOSSology
-  Upload a file for scanning.
-  Generate a CycloneDX report from FOSSology UI.

cc: @GMishx @shaheemazmalmmd @avinal @hastagAB @Kaushl2208 

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2507"></a>

### CODE REVIEW

File: src/cyclonedx/agent/cyclonedx.php [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+      $filesWithLicenses[$finding['uploadtree_pk']]['copyrights'][] = \convertToUTF8($finding['textfinding'],false);
+    }
+  }
+
+  /**
+   * @brief Generate the components by files
+   * @param string[][][] $filesWithLicenses
+   * @param string $treeTableName
+   * @param int $uploadId
+   * @return string Components string
```
Comment (GMishx) [inline src/cyclonedx/agent/cyclonedx.php:379]:
```suggestion
   * @return array Components list
```

File: src/cyclonedx/agent/cyclonedx.php [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+      $licensedata['id'] = $spdxId;
+      $licensedata['url'] = $licenseTexts[$spdxId]['url'];
+      $mainLicenses[] = $this->reportGenerator->createLicense($licensedata);
+    }
+
+    $hashes = $this->uploadDao->getUploadHashes($uploadId);
+    $serializedhash = array();
+    $serializedhash[] = $this->reportGenerator->createHash('SHA-1', $hashes['sha1']);
+    $serializedhash[] = $this->reportGenerator->createHash('MD5', $hashes['md5']);
+    $serializedhash[] = $this->reportGenerator->createHash('SHA-256', $hashes['sha256']);
```
Comment (GMishx) [inline src/cyclonedx/agent/cyclonedx.php:203]: Please check if `sha256` is not empty before processing it as old uploads might not contain that value.

File: src/cyclonedx/agent/cyclonedx.php [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+    $maincomponentData = array (
+      'bomref' => strval($uploadId),
+      'type' => 'library',
+      'name' => $upload->getFilename(),
+      'hashes' => $serializedhash,
+      'scope' => 'required',
+      'mimeType' => 'text/plain',
```
Comment (GMishx) [inline src/cyclonedx/agent/cyclonedx.php:211]: mimeType of a main package cannot be `text/plain` :smiley:

File: src/cyclonedx/agent/reportgenerator.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  /**
+   * Generates the report.
+   *
+   * @param array $bomData The BOM data.
+   * @return array The generated report.
+   */
+  public function generateReport($bomdata): array
+  {
+    $report = [
+      'bomFormat' => 'CycloneDX',
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:59]: Please add schema also `"$schema": "http://cyclonedx.org/schema/bom-1.4.schema.json"`

File: src/cyclonedx/agent/reportgenerator.php [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+      'type' => $componentData['type'],
+      'name' => $componentData['name']
+    ];
+
+    if (array_key_exists('mimeType', $componentData) && !empty($componentData['mimeType'])) {
+      $component['mime-type'] = $componentData['mimeType'];
+    }
+
+    if (array_key_exists('bomref', $componentData) && !empty($componentData['bomref'])) {
+      $component['bomref'] = $componentData['bomref'];
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:98]:
```suggestion
      $component['bom-ref'] = $componentData['bomref'];
```

File: src/cyclonedx/ui/CycloneDXGeneratorUi.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+use Fossology\Lib\Dao\FolderDao;
+use Fossology\Lib\Dao\UploadDao;
+use Fossology\Lib\Data\Upload\Upload;
+use Fossology\Lib\Plugin\DefaultPlugin;
+use Symfony\Component\HttpFoundation\Request;
+
+class CycloneDXGeneratorUi extends DefaultPlugin
+{
+  const NAME = 'ui_cyclonedx';
+  const DEFAULT_OUTPUT_FORMAT = "cyclonedx";
```
Comment (GMishx) [inline src/cyclonedx/ui/CycloneDXGeneratorUi.php:21]: If updating output format in `cyclonedx.php`, please remember to update it here as well.

File: src/cyclonedx/agent/reportgenerator.php [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    /**
+     * "Specifies the scope of the component. If scope is not specified,
+     *  'required' scope SHOULD be assumed by the consumer of the BOM."
+     */
+    if (array_key_exists('scope', $componentData) && !empty($componentData['scope'])) {
+      $component['scope'] = $componentData['scope'];
+    }
+    else {
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:108]:
```suggestion
    } else {
```

File: src/cyclonedx/agent/reportgenerator.php [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+   * @return array The generated license.
+   */
+  private function generateLicense(array $licenseData): array
+  {
+    $license = [];
+
+    if (array_key_exists('id', $licenseData) && !empty($licenseData['id'])) {
+      $license['license']['id'] = $licenseData['id'];
+    }
+    else if (array_key_exists('name', $licenseData) && !empty($licenseData['name'])) {
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:144]:
```suggestion
    } else if (array_key_exists('name', $licenseData) && !empty($licenseData['name'])) {
```

File: src/cyclonedx/agent/cyclonedx.php [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+            $licensedata = array(
+              "id"   => $licenseTexts[$license]["id"],
+              "name" => $licenseTexts[$license]["name"],
+              "url"  => $licenseTexts[$license]["url"]
+            );
+            $licensesfound[] = $this->reportGenerator->createLicense($licensedata);
+          }
+        }
+      }
+      else {
```
Comment (GMishx) [inline src/cyclonedx/agent/cyclonedx.php:427]:
```suggestion
      } else {
```

**Review by GMishx — CHANGES_REQUESTED** [review]

File: src/cyclonedx/cyclonedx.conf [hunk 10]
```diff
@@ -0,0 +1,23 @@
+; SPDX-FileCopyrightText: © 2023 Sushant Kumar <sushantmishra02102002@gmail.com>
+
+; SPDX-License-Identifier: FSFAP
```
[the remark this answers is not in the document]
Comment (GMishx) [reply src/cyclonedx/cyclonedx.conf:3]: It is not an actual code, so FSFAP is an appropriate license here.

[the remark this answers is not in the document]
Comment (avinal) [reply src/cyclonedx/cyclonedx.conf:3]: Got it, thanks.

File: src/cyclonedx/agent/cyclonedx.php [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+include_once(__DIR__ . "/reportgenerator.php");
+
+/**
+ * @class cyclonedxAgent
+ * @brief cyclonedxAgent agent generates SBOM in cyclonedx format
+ */
+class CycloneDXAgent extends Agent
+{
+  const OUTPUT_FORMAT_KEY = "outputFormat";               ///< Argument key for output format
+  const DEFAULT_OUTPUT_FORMAT = "cyclonedx";                  ///< Default output format
```
[the remark this answers is not in the document]
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Done

File: src/cyclonedx/agent/cyclonedx.php [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+    $this->reportGenerator = new BomReportGenerator();
+  }
+
+  /**
+   * @copydoc Fossology::Lib::Agent::Agent::processUploadId()
+   * @see Fossology::Lib::Agent::Agent::processUploadId()
+   */
+  function processUploadId($uploadId)
+  {
+    global $SysConf;
```
[the remark this answers is not in the document]
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Done

File: src/cyclonedx/agent/cyclonedx.php [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+    $this->licenseMap = new LicenseMap($this->dbManager, $this->groupId, LicenseMap::REPORT, true);
+
+    $packageNodes = $this->renderPackage($uploadId);
+
+    $this->computeUri($uploadId);
+
+    $contents = json_encode($packageNodes, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+
+    $this->writeReport($contents, $uploadId);
```
[the remark this answers is not in the document]
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Done

File: src/cyclonedx/agent/cyclonedx.php [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+  /**
+   * @brief Get the URI for the given package
+   * @param string $fileBase Name of the upload
+   * @return string URI for the upload
+   */
+  protected function getUri($fileBase)
+  {
+    $fileName = $fileBase. strtoupper($this->outputFormat)."_".$this->packageName.'_'.date("Y-m-d_H:i:s");
+    $fileName = $fileName .".json" ;
+    return $fileName;
```
[the remark this answers is not in the document]
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Done

File: src/cyclonedx/agent/cyclonedx.php (hunk 3 above)
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Yup, I have changed that.

File: src/cyclonedx/agent/cyclonedx.php [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+    }
+  }
+
+  /**
+   * @brief Attach finding agents to the files and return names of scanners
+   * @param[in,out] string &$filesWithLicenses
+   * @param ItemTreeBounds $itemTreeBounds
+   * @return string Name(s) of scanners used
+   */
+  protected function addScannerResults(&$filesWithLicenses, ItemTreeBounds $itemTreeBounds)
```
[the remark this answers is not in the document]
Comment (GMishx) [reply src/cyclonedx/agent/cyclonedx.php:296]: A gentle reminder.

File: src/cyclonedx/agent/reportgenerator.php [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+   * Generates the report.
+   *
+   * @param array $bomData The BOM data.
+   * @return array The generated report.
+   */
+  public function generateReport($bomdata): array
+  {
+    $report = [
+      'bomFormat' => 'CycloneDX',
+      'schema' => 'http://cyclonedx.org/schema/bom-1.4.schema.json',
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:60]:
```suggestion
      '$schema' => 'http://cyclonedx.org/schema/bom-1.4.schema.json',
```

File: src/cyclonedx/agent/reportgenerator.php [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+  /**
+   * Generates a license.
+   *
+   * @param array $licenseData The license data.
+   * @return array The generated license.
+   */
+  private function generateLicense(array $licenseData): array
+  {
+    $license = [];
+
```
Comment (GMishx) [inline src/cyclonedx/agent/reportgenerator.php:140]: CycloneDX schema does not support SPDX LicenseRef as valid license id. Proposing following changes to make LicenseRefs as a license expression.

Comment (GMishx) [reply src/cyclonedx/agent/reportgenerator.php:140]:
```diff
diff --git a/src/cyclonedx/agent/reportgenerator.php b/src/cyclonedx/agent/reportgenerator.php
index 90555bb4a..47d60049e 100644
--- a/src/cyclonedx/agent/reportgenerator.php
+++ b/src/cyclonedx/agent/reportgenerator.php
@@ -11,6 +11,8 @@
  */
 namespace Fossology\CycloneDX;
 
+use Fossology\Lib\Data\LicenseRef;
+
 class BomReportGenerator
 {
   /**
@@ -138,6 +140,13 @@ class BomReportGenerator
   {
     $license = [];
 
+    // Check license ID is a LicenseRef
+    if (array_key_exists('id', $licenseData) && !empty($licenseData['id']) &&
+      stripos($licenseData['id'], LicenseRef::SPDXREF_PREFIX) === 0) {
+      $license['expression'] = $licenseData['id'];
+      return $license;
+    }
+
     if (array_key_exists('id', $licenseData) && !empty($licenseData['id'])) {
       $license['license']['id'] = $licenseData['id'];
     } else if (array_key_exists('name', $licenseData) && !empty($licenseData['name'])) {
```

**Review by GMishx — CHANGES_REQUESTED** [review]

**Comment by GMishx** [comment]
Also, please do not forgot to work on https://github.com/fossology/fossology/pull/2507#discussion_r1257742021

File: src/cyclonedx/agent/reportgenerator.php (hunk 16 above)
Comment (its-sushant) [reply src/cyclonedx/agent/reportgenerator.php]: Done

File: src/cyclonedx/agent/reportgenerator.php (hunk 17 above)
Comment (its-sushant) [reply src/cyclonedx/agent/reportgenerator.php:142]: Done

File: src/cyclonedx/agent/cyclonedx.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (its-sushant) [reply src/cyclonedx/agent/cyclonedx.php]: Done

**Comment by its-sushant** [comment]
hello, I have made the changes as per suggestion. Also, refactored SPDX agent code.

cc: @GMishx @avinal

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Comment by GMishx** [comment]
@its-sushant , please use the following patch to fix the SPDX report agent and its test cases.
And please make sure to apply all the changes suggested bellow.

```diff
diff --git a/src/cyclonedx/agent/cyclonedx.php b/src/cyclonedx/agent/cyclonedx.php
index 9e6f224dc..9400e7949 100755
--- a/src/cyclonedx/agent/cyclonedx.php
+++ b/src/cyclonedx/agent/cyclonedx.php
@@ -28,7 +28,6 @@ use Fossology\Lib\Data\AgentRef;
 use Fossology\Lib\Data\DecisionTypes;
 use Fossology\Lib\Data\LicenseRef;
 use Fossology\Lib\Data\Tree\ItemTreeBounds;
-use Fossology\Lib\Data\Upload\Upload;
 use Fossology\Lib\Db\DbManager;
 use Fossology\Lib\Report\ReportUtils;
 
@@ -168,7 +167,7 @@ class CycloneDXAgent extends Agent
     $this->reportutils->addClearingStatus($filesWithLicenses, $itemTreeBounds, $this->groupId);
     $this->heartbeat(0);
 
-    $this->reportutils->addScannerResults($filesWithLicenses, $itemTreeBounds, $this->groupId);
+    $this->reportutils->addScannerResults($filesWithLicenses, $itemTreeBounds, $this->groupId, $this->includedLicenseIds);
     $this->heartbeat(0);
 
     $this->reportutils->addCopyrightResults($filesWithLicenses, $uploadId);
@@ -179,7 +178,7 @@ class CycloneDXAgent extends Agent
 
     $mainLicenseIds = $this->clearingDao->getMainLicenseIds($uploadId, $this->groupId);
     $mainLicenses = array();
-    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId);
+    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId, $this->includedLicenseIds);
     foreach ($mainLicenseIds as $licId) {
       $reportedLicenseId = $this->licenseMap->getProjectedId($licId);
       $this->includedLicenseIds[$reportedLicenseId] = true;
@@ -276,7 +275,7 @@ class CycloneDXAgent extends Agent
     $filesProceeded = 0;
     $lastValue = 0;
     $components = array();
-    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId);
+    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId, $this->includedLicenseIds);
     foreach ($filesWithLicenses as $fileId=>$licenses) {
       $filesProceeded += 1;
       if (($filesProceeded & 2047) == 0) {
diff --git a/src/lib/php/Report/ReportUtils.php b/src/lib/php/Report/ReportUtils.php
index c8a6f8e51..4e5ed9167 100644
--- a/src/lib/php/Report/ReportUtils.php
+++ b/src/lib/php/Report/ReportUtils.php
@@ -6,13 +6,13 @@
 */
 namespace Fossology\Lib\Report;
 
+use Fossology\Lib\BusinessRules\LicenseMap;
 use Fossology\Lib\Data\AgentRef;
-use Fossology\Lib\Proxy\UploadTreeProxy;
-use Fossology\Lib\Proxy\ScanJobProxy;
-use Fossology\Lib\Proxy\LicenseViewProxy;
 use Fossology\Lib\Data\LicenseRef;
 use Fossology\Lib\Data\Tree\ItemTreeBounds;
-use Fossology\Lib\BusinessRules\LicenseMap;
+use Fossology\Lib\Proxy\LicenseViewProxy;
+use Fossology\Lib\Proxy\ScanJobProxy;
+use Fossology\Lib\Proxy\UploadTreeProxy;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 
 
@@ -39,12 +39,14 @@ class ReportUtils
    */
   private $uploadDao;
 
-  function __construct() {
+  function __construct()
+  {
     global $container;
     $this->container = $container;
 
     $this->dbManager = $this->container->get('db.manager');
     $this->uploadDao = $this->container->get('dao.upload');
+    $this->licenseMap = null;
   }
 
   /**
@@ -67,20 +69,24 @@ class ReportUtils
     $alreadyClearedUploadTreeView->unmaterialize();
 
     $uploadTreeIds = array_keys($filesWithLicenses);
-    foreach ($uploadTreeIds as $uploadTreeId)
-      $filesWithLicenses[$uploadTreeId]['isCleared'] = false === array_key_exists($uploadTreeId,$filesThatShouldStillBeCleared);
+    foreach ($uploadTreeIds as $uploadTreeId) {
+      $filesWithLicenses[$uploadTreeId]['isCleared'] = false === array_key_exists($uploadTreeId, $filesThatShouldStillBeCleared);
+    }
   }
 
-    /**
+  /**
    * @brief Attach finding agents to the files and return names of scanners
    * @param[in,out] string &$filesWithLicenses
    * @param ItemTreeBounds $itemTreeBounds
    * @param int $groupId
-   * @return string Name(s) of scanners used
+   * @param[in,out] bool &$includedLicenseIds
+   * @return array Name(s) of scanners used
    */
-  public function addScannerResults(&$filesWithLicenses, ItemTreeBounds $itemTreeBounds, $groupId)
+  public function addScannerResults(&$filesWithLicenses, ItemTreeBounds $itemTreeBounds, $groupId, &$includedLicenseIds)
   {
-    $this->licenseMap = new LicenseMap($this->dbManager, $groupId, LicenseMap::REPORT, true);
+    if ($this->licenseMap === null) {
+      $this->licenseMap = new LicenseMap($this->dbManager, $groupId, LicenseMap::REPORT, true);
+    }
     $uploadId = $itemTreeBounds->getUploadId();
     $scannerAgents = array_keys($this->agentNames);
     $scanJobProxy = new ScanJobProxy($this->container->get('dao.agent'), $uploadId);
@@ -112,14 +118,14 @@ class ReportUtils
         } else {
           $filesWithLicenses[$row['uploadtree_pk']]['scanner'][] = "";
         }
-        $this->includedLicenseIds[$reportedLicenseId] = true;
+        $includedLicenseIds[$reportedLicenseId] = true;
       }
     }
     $this->dbManager->freeResult($res);
     return $scannerIds;
   }
 
-    /**
+  /**
    * @brief Add copyright results to the files
    * @param[in,out] string &$filesWithLicenses
    * @param int $uploadId
@@ -156,12 +162,13 @@ class ReportUtils
     }
   }
 
-    /**
+  /**
    * @brief Get the license texts from fossology
    * @param int $groupId
+   * @param[in,out] bool &$includedLicenseIds
    * @return string[] with keys being shortname
    */
-  public function getLicenseTexts($groupId)
+  public function getLicenseTexts($groupId, &$includedLicenseIds)
   {
     $licenseTexts = array();
     $licenseViewProxy = new LicenseViewProxy($groupId,
@@ -172,7 +179,7 @@ class ReportUtils
     $res = $this->dbManager->execute($stmt);
 
     while ($row = $this->dbManager->fetchArray($res)) {
-      if (array_key_exists($row['rf_pk'], $this->includedLicenseIds)) {
+      if (array_key_exists($row['rf_pk'], $includedLicenseIds)) {
         $shortname = $row['rf_shortname'];
         $spdxId = LicenseRef::convertToSpdxId($shortname, $row['rf_spdx_id']);
         $licenseTexts[$spdxId] = [
@@ -183,7 +190,7 @@ class ReportUtils
         ];
       }
     }
-    foreach ($this->includedLicenseIds as $license => $customText) {
+    foreach ($includedLicenseIds as $license => $customText) {
       if (true !== $customText) {
         $licenseTexts[$license] = [
           'text' => $customText,
@@ -196,4 +203,4 @@ class ReportUtils
     $this->dbManager->freeResult($res);
     return $licenseTexts;
   }
-}
\ No newline at end of file
+}
diff --git a/src/spdx2/agent/spdx2.php b/src/spdx2/agent/spdx2.php
index deece4a72..ce2e06e47 100644
--- a/src/spdx2/agent/spdx2.php
+++ b/src/spdx2/agent/spdx2.php
@@ -323,9 +323,10 @@ class SpdxTwoAgent extends Agent
     $this->reportutils->addClearingStatus($filesWithLicenses,$itemTreeBounds, $this->groupId);
     $this->heartbeat(0);
 
-    $scannerIDs = $this->reportutils->addScannerResults($filesWithLicenses, $itemTreeBounds, $this->groupId);
+    $scannerIDs = $this->reportutils->addScannerResults($filesWithLicenses, $itemTreeBounds, $this->groupId,
+      $this->includedLicenseIds);
     $licenseComment = "";
-    if(!empty($scannerIDs)) {
+    if (!empty($scannerIDs)) {
       $licenseComment = $this->getLicenseComment($scannerIDs);
     }
     $this->heartbeat(0);
@@ -569,7 +570,7 @@ class SpdxTwoAgent extends Agent
     }
     umask(0133);
 
-    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId);
+    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId, $this->includedLicenseIds);
     if (strcmp($this->outputFormat, "dep5")!==0) {
       $licenseTexts = SpdxTwoUtils::addPrefixOnDemandKeys($licenseTexts);
     }
@@ -652,7 +653,7 @@ class SpdxTwoAgent extends Agent
     $filesProceeded = 0;
     $lastValue = 0;
     $content = '';
-    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId);
+    $licenseTexts = $this->reportutils->getLicenseTexts($this->groupId, $this->includedLicenseIds);
     if (! in_array(self::DATA_LICENSE, $this->licenseTextPrinted)) {
       $this->licenseTextPrinted[] = self::DATA_LICENSE;
     }
```

**Comment by its-sushant** [comment]
Done, @GMishx

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Comment by shaheemazmalmmd** [comment]
Not able to download the report via UI. @its-sushant can you add cycloneDx report in AjaxShowJobs.php so that it can be downloaded ??

screenShot81

**Comment by its-sushant** [comment]
> Not able to download the report via UI. @its-sushant can you add cycloneDx report in AjaxShowJobs.php so that it can be downloaded ??
> 
> screenShot81

Hello @shaheemazmalmmd sir, I have added the option to download report from UI.

**Pull request merged**